src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java

```diff
@@ -0,0 +1,505 @@
+        } catch (Exception e) {
+            LOGGER.error(
+                    "Reprocess cleaner encountered an exception {}:\n{}\n{}",
+                    e.getClass(),
+                    e.getMessage(),
+                    e.getStackTrace());
+        }
+    }
+
+    public static class FileCategorizer {
+        private final Set<String> dirtyFiles = new HashSet<>();
+        private final Map<String, IngestEntry> stageFiles = new HashMap<>();
+        private final long currentOffset;
+
+        static FileCategorizer build(Collection<String> files, long currentOffset) {
+            return files
+                    .stream()
+                    .reduce(
+                        new FileCategorizer(currentOffset),
+                        FileCategorizer::categorizeFile,
+                        FileCategorizer::combine);
+        }
+
+        private FileCategorizer(long startOffset) {
+            this.currentOffset = startOffset;
+        }
+
+        private static FileCategorizer combine(FileCategorizer lhs, FileCategorizer rhs) {
+            FileCategorizer result = new FileCategorizer(lhs.currentOffset);
+            result.stageFiles.putAll(lhs.stageFiles);
+            result.stageFiles.putAll(rhs.stageFiles);
+            result.dirtyFiles.addAll(lhs.dirtyFiles);
+            result.dirtyFiles.addAll(rhs.dirtyFiles);
+            return result;
+        }
+
+        private FileCategorizer categorizeFile(String file) {
+            long fileOffset = FileNameUtils.fileNameToStartOffset(file);
+            long timestamp = FileNameUtils.fileNameToTimeIngested(file);
+            if (currentOffset <= fileOffset) {
+                dirtyFiles.add(file);
+            } else {
+                IngestEntry entry = new IngestEntry(InternalUtils.IngestedFileStatus.NOT_FOUND, timestamp);
+                stageFiles.put(file, entry);
+            }
+            return this;
+        }
+
+        void updateFileStatus(Map<String, InternalUtils.IngestedFileStatus> report) {
+            report.forEach((fileName, status) -> stageFiles.computeIfPresent(fileName, (key, entry) -> {
+                entry.status = status;
+                return entry;
+            }));
+        }
+
+        boolean hasDirtyFiles() {
+            return !dirtyFiles.isEmpty();
+        }
+
+        boolean hasStageFiles() {
+            return !stageFiles.isEmpty();
+        }
+
+        Stream<String> query(Predicate<Map.Entry<String, IngestEntry>> filter) {
+            return stageFiles
+                    .entrySet()
+                    .stream()
+                    .filter(filter)
+                    .map(Map.Entry::getKey);
+        }
+
+        void stopTrackingFiles(List<String> files) {
+            files.forEach(file -> stageFiles.computeIfPresent(file, (key, entry) -> {
+                entry.keepTracking = false;
+                return entry;
+            }));
+        }
+    }
+
+    @VisibleForTesting
+    static class IngestEntry {
```

Comment:
It isn't clear what `IngestEntry` really means. Let's name it `IngestedFileState` or somehow similar to it.